### PR TITLE
Expose APIs for rz-solver

### PR DESCRIPTION
**Pull request by giridharprasath**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Exposing some APIs required for rz-solver
#4563 

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/core/cmd/cmd_search_rop.c [hunk 1]
```diff
@@ -289,7 +289,7 @@ static bool parse_reg_op_const(const RzCore *core, const char *str, RzRopConstra
 RZ_API RZ_OWN RzRopSearchContext *rz_core_rop_search_context_new(RZ_NONNULL const RzCore *core, RZ_NULLABLE const char *greparg, const bool regexp,
 	const RzRopRequestMask mask, RZ_NONNULL RZ_BORROW RzCmdStateOutput *state) {
 
-	rz_return_val_if_fail(core && state, NULL);
+	rz_return_val_if_fail(core, NULL);
```
Comment (XVilka) [inline librz/core/cmd/cmd_search_rop.c:445]: Why? It says the sate also shouldn't be NULL

File: librz/core/rop.c [hunk 2]
```diff
@@ -1045,12 +1041,12 @@ static bool print_rop(const RzCore *core, RzList /*<RzCoreAsmHit *>*/ *hitlist,
 		break;
 		// fallthrough
```
Comment (XVilka) [inline librz/core/rop.c]: It's not a fallthrough btw, since there is a `break` before

**Review by XVilka — CHANGES_REQUESTED** [review]
I recommend writing a unit test for these ROP APIs, it will certainly help with the development and automated testing.

File: librz/core/cmd/cmd_search_rop.c (hunk 1 above)
Comment (giridharprasath) [reply librz/core/cmd/cmd_search_rop.c:445]: Noted :+1:  Changed it to `RZ_NULLABLE`, as the rz-solver can [pass](https://github.com/giridharprasath/rz-solver/blob/186bf295d2d29f62696a3f9946c61850f598b677/tests/test_rop_solver.c#L106) it as NULL

**Comment by giridharprasath** [comment]
> I recommend writing a unit test for these ROP APIs, it will certainly help with the development and automated testing.

Agree :+1: , Will track it under another issue

**Comment by giridharprasath** [comment]
APIs used by the rz-solver: [test-rop-solver.c](https://github.com/giridharprasath/rz-solver/blob/dev/tests/test_rop_solver.c)

**Comment by XVilka** [comment]
@giridharprasath ping?

**Comment by XVilka** [comment]
@giridharprasath please rebase and add a unit test for this. Then we can merge.

**Comment by giridharprasath** [comment]
@XVilka I will be wrapping this and push the changes today.

**Comment by giridharprasath** [comment]
> @XVilka I will be wrapping this and push the changes today.

Got a little late, Will commit the changes today

File: librz/include/rz_util/rz_str_util.h [hunk 3]
```diff
@@ -6,6 +6,17 @@
...
+	(((x) == '+' && (y) == '+') || \
+		((x) == '-' && (y) == '-') || \
+		((x) == '+' && (y) == '=') || \
+		((x) == '-' && (y) == '=') || \
+		((x) == '*' && (y) == '=') || \
+		((x) == '/' && (y) == '=') || \
+		((x) == '%' && (y) == '=') || \
+		((x) == '&' && (y) == '=') || \
+		((x) == '|' && (y) == '=') || \
+		((x) == '^' && (y) == '='))
```
Comment (wargio) [inline librz/include/rz_util/rz_str_util.h]: i'm not sure if this is the correct location. i would define this in the rop headers? maybe call it RZ_ROP_IS_COMPOUND_OP

Comment (XVilka) [reply librz/include/rz_util/rz_str_util.h]: Agree, it also could be an inline function instead.

Comment (wargio) [reply librz/include/rz_util/rz_str_util.h]: that is a good idea. an inline function is much better but do not add an inline function in the headers.

Comment (giridharprasath) [reply librz/include/rz_util/rz_str_util.h]: Agreed :+1: , I have updated it as a `static inline` function

**Review by XVilka — APPROVED** [review]

File: librz/core/cmd/cmd_search_rop.c [hunk 4]
```diff
@@ -10,11 +10,27 @@
 #include "rz_rop.h"
 
 static void skip_whitespace(const char *str, ut64 *idx) {
+	if (*idx >= strlen(str)) {
```
Comment (wargio) [inline librz/core/cmd/cmd_search_rop.c:13]: is there any possibility that `str` is NULL? if is not possible, ignore this.

Comment (giridharprasath) [reply librz/core/cmd/cmd_search_rop.c:13]: This is always called from this [code](https://github.com/giridharprasath/rizin/blob/3aeef0a25afd7a4747c86d1f9d262722d2deb1d2/librz/core/cmd/cmd_search_rop.c#L565) path

**Review by wargio — APPROVED** [review]

**Pull request merged**